### PR TITLE
Fix BST entry preview formatting

### LINKED ISSUE #16330: Some advanced formatting does not render in BST Entry Preview
### JabRef version

6.0-alpha.6

### Operating system

Windows

### Details on version and operating system

Windows 11

### Checked with the [latest development build](https://builds.jabref.org/main/) (copy version output from About dialog)

- [x] I made a backup of my libraries before testing the latest development version.
- [x] I have tested the latest development version and the problem persists

### Steps to reproduce the behaviour

Follow-up issue to https://github.com/JabRef/jabref/issues/16314
Certain advanced formatting options such as smallcaps, subscript, superscript are not supported by our BST Entry Preview.

For example, on importing [jfm.bst](https://github.com/jeanluct/braidlab/blob/master/doc/jfm.bst) and trying to preview the Cooper et al entry from Chocolate.bib:

Image

Even with the fixes of https://github.com/JabRef/jabref/pull/16315.

The entry preview should have the authors in smallcaps and ideally render as:

Image

(taken from OO Bibliography rendering as per https://github.com/JabRef/jabref/pull/16315 - which is fine).

The disparity may be due to rich html renderer of the entry preview not being able to parse these special tags that specify these styles.

### Appendix

...
<details>

<summary>Log File</summary>

```
Paste an excerpt of your log file here
```

</details>

### PR DESCRIPTION
**Pull request by subhramit**
### Summary
Fix BST preview formatting for smallcaps.

### Steps to test

1. Open example library
2. View entry preview of any entry using [`jfm.bst`](https://github.com/jeanluct/braidlab/blob/master/doc/jfm.bst)
3. Observe that author is rendered in smallcaps properly

image
image

### Related issues and pull requests

Closes https://github.com/JabRef/jabref/issues/16330

### AI usage

Zed, GPT 5.4

### Checklist

- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] If AI tools were used, I disclosed them in the "AI usage" section and reviewed, understood, and take full ownership of all AI-generated code
- [x] I manually tested my changes in running JabRef (always required)
- [x] I added JUnit tests for changes (if applicable)
- [x] I added screenshots in the PR description (if change is visible to the user)
- [x] I added **one sentence (max 20 words)** to `CHANGELOG.md` describing the change from the user's point of view (if the change is visible to the user)
- [/] I checked the [user documentation](https://docs.jabref.org/) for up to dateness and submitted a pull request to our [user documentation repository](https://github.com/JabRef/user-documentation/tree/main/en)

### CODE REVIEW

**Comment by calixtus** [comment]
Missing tests for edge cases and malformed strings

File: jablib/src/main/java/org/jabref/logic/preview/BstPreviewLayout.java [hunk 1]
```diff
@@ -175,6 +173,124 @@ public String generatePreview(BibEntry originalEntry, BibDatabaseContext databas
         return result;
     }
 
+    private static String convertInlineSmallCapsToHtml(String latex) {
+        String html = replaceCommandWithBalancedArgument(latex, "textsc");
+        return replaceLegacySwitchWithBalancedArgument(html, "sc");
+    }
+
+    private static String replaceCommandWithBalancedArgument(String input, String command) {
+        String needle = "\\" + command;
```
Comment (pluto-han) [inline jablib/src/main/java/org/jabref/logic/preview/BstPreviewLayout.java]: Im not sure about the name "needle"?

File: jablib/src/test/java/org/jabref/logic/bst/BstPreviewLayoutTest.java [hunk 2]
```diff
@@ -62,6 +62,51 @@ void mathSymbolsInBracedMathAreConvertedToUnicode() throws URISyntaxException {
...
     }
 
+    @Test
+    void smallCapsFormattingIsRendered() throws URISyntaxException {
+        BstPreviewLayout bstPreviewLayout = new BstPreviewLayout(Path.of(BstPreviewLayoutTest.class.getResource("abbrv.bst").toURI()));
+        BibEntry entry = new BibEntry().withField(StandardField.AUTHOR, "Oliver Kopp")
+                                       .withField(StandardField.TITLE, "\\textsc{L{\\'o}pez}");
+        String preview = bstPreviewLayout.generatePreview(entry, bibDatabaseContext);
+        assertEquals("O.\u00a0Kopp. <span style=\"font-variant: small-caps\">López.</span>", preview);
+    }
```
Comment (pluto-han) [inline jablib/src/test/java/org/jabref/logic/bst/BstPreviewLayoutTest.java]: all the tests use the same pattern, i think we can extract them and use @ParameterizedTest?

**Comment by subhramit** [comment]
> Missing tests for edge cases and malformed strings

Added

**Review by pluto-han — COMMENTED** [review]
nitpick

File: jablib/src/test/java/org/jabref/logic/bst/BstPreviewLayoutTest.java (hunk 2 above)
Comment (subhramit) [reply jablib/src/test/java/org/jabref/logic/bst/BstPreviewLayoutTest.java]: I am a little torn between having easy-to-understand cases via method names vs parameterizing here.
Perhaps parameterizing is cleaner as LLMs can quickly pinpoint failures today.
Doing

Comment (pluto-han) [reply jablib/src/test/java/org/jabref/logic/bst/BstPreviewLayoutTest.java]: > I am a little torn between having easy-to-understand cases via method names vs parameterizing here. Perhaps parameterizing is cleaner as LLMs can quickly pinpoint failures today. Doing

Adding comments on top of each test can help maybe

File: jablib/src/main/java/org/jabref/logic/preview/BstPreviewLayout.java (hunk 1 above)
Comment (subhramit) [reply jablib/src/main/java/org/jabref/logic/preview/BstPreviewLayout.java]: I first used it in https://github.com/JabRef/jabref/blob/b52e3d22841c5022aede529d785397ad1a1f2dd9/jablib/src/main/java/org/jabref/logic/openoffice/bst/BSTFormatUtils.java#L80

Context: `needle` is standard “needle in a haystack” terminology for substring search, `input` being the haystack, `needle` being the susbtrign we are looking for.

But it is possible that someone reading is unaware, so I'll rename everywhere to not cause extra contextual load.

File: jablib/src/test/java/org/jabref/logic/bst/BstPreviewLayoutTest.java (hunk 2 above)
Comment (subhramit) [reply jablib/src/test/java/org/jabref/logic/bst/BstPreviewLayoutTest.java]: > Adding comments on top of each test can help maybe

Yes good idea, `CSLFormatUtils` does that. Though I was talking more about getting the failure context ("what" failed) during the test failure result, where comments cannot be surfaced. But junit has numbering as well, so should be fine.

Comment (subhramit) [reply jablib/src/test/java/org/jabref/logic/bst/BstPreviewLayoutTest.java]: Also this was really a good cleanup, not a nitpick :)

Comment (pluto-han) [reply jablib/src/test/java/org/jabref/logic/bst/BstPreviewLayoutTest.java]: It looks much better, thanks!

**Comment by Siedlerchr** [comment]
looks good on my machine

**Review by Siedlerchr — APPROVED** [review]
